clippy_lints/src/map_nil_fn.rs

```diff
@@ -0,0 +1,187 @@
+use rustc::hir;
+use rustc::lint::*;
+use rustc::ty;
+use syntax::codemap::Span;
+use utils::{in_macro, iter_input_pats, match_type, method_chain_args, snippet, span_lint_and_then};
+use utils::paths;
+
+#[derive(Clone)]
+pub struct Pass;
+
+/// **What it does:** Checks for usage of `Option.map(f)` where f is a nil
+/// function or closure
+///
+/// **Why is this bad?** Readability, this can be written more clearly with
+/// an if statement
+///
+/// **Known problems:** None.
+///
+/// **Example:**
+/// ```rust
+/// let x : Option<&str> = do_stuff();
+/// x.map(log_err_msg);
+/// x.map(|msg| log_err_msg(format_msg(msg)))
+/// ```
+/// The correct use would be:
+/// ```rust
+/// let x : Option<&str> = do_stuff();
+/// if let Some(msg) = x {
+///     log_err_msg(msg)
+/// }
+/// if let Some(msg) = x {
+///     log_err_msg(format_msg(msg))
+/// }
+/// ```
+declare_lint! {
+    pub OPTION_MAP_NIL_FN,
+    Allow,
+    "using `Option.map(f)`, where f is a nil function or closure"
+}
+
+
+impl LintPass for Pass {
+    fn get_lints(&self) -> LintArray {
+        lint_array!(OPTION_MAP_NIL_FN)
+    }
+}
+
+fn is_nil_type(ty: ty::Ty) -> bool {
+    match ty.sty {
+        ty::TyTuple(slice) => slice.is_empty(),
+        ty::TyNever => true,
+        _ => false,
+    }
+}
+
+fn is_nil_function(cx: &LateContext, expr: &hir::Expr) -> bool {
+    let ty = cx.tables.expr_ty(expr);
+
+    if let ty::TyFnDef(_, _, bare) = ty.sty {
+        if let Some(fn_type) = cx.tcx.no_late_bound_regions(&bare.sig) {
+            return is_nil_type(fn_type.output());
+        }
+    }
+    false
+}
+
+fn is_nil_expression(cx: &LateContext, expr: &hir::Expr) -> bool {
+    is_nil_type(cx.tables.expr_ty(expr))
+}
+
+// The expression inside a closure may or may not have surrounding braces and
+// semicolons, which causes problems when generating a suggestion. Given an
+// expression that evaluates to '()' or '!', recursively remove useless braces
+// and semi-colons until is suitable for including in the suggestion template
+fn reduce_nil_expression<'a>(cx: &LateContext, expr: &'a hir::Expr) -> Option<Span> {
+    if !is_nil_expression(cx, expr) {
```

Comment:
I wonder if the bulk of this function could be moved into utils as a general-purpose closure-reducing function.